src/liboslexec/shadingsys.cpp

```diff
@@ -3710,67 +3710,46 @@ ShadingSystemImpl::merge_instances (ShaderGroup &group, bool post_opt)
 
 
 
-#if OIIO_HAS_COLORPROCESSOR
+#ifndef __CUDACC__
 
 OIIO::ColorProcessorHandle
-OCIOColorSystem::load_transform (StringParam fromspace, StringParam tospace)
+OCIOColorSystem::load_transform (StringParam fromspace, StringParam tospace,
+                                 ShadingSystemImpl* ss)
 {
     if (fromspace != m_last_colorproc_fromspace ||
         tospace != m_last_colorproc_tospace) {
-        m_last_colorproc = m_colorconfig.createColorProcessor (fromspace, tospace);
+        m_last_colorproc = colorconfig(ss).createColorProcessor(fromspace, tospace);
         m_last_colorproc_fromspace = fromspace;
         m_last_colorproc_tospace = tospace;
     }
     return m_last_colorproc;
 }
 
-#endif
-
 
 
-template <> bool
-ShadingSystemImpl::ocio_transform (StringParam fromspace, StringParam tospace,
-                                   const Color3& C, Color3& Cout) {
-#if OIIO_HAS_COLORPROCESSOR
-    OIIO::ColorProcessorHandle cp;
-    {
-        lock_guard lock (m_mutex);
-        cp = m_ocio_system.load_transform(fromspace, tospace);
-    }
-    if (cp) {
-        Cout = C;
-        cp->apply ((float *)&Cout);
-        return true;
+const OIIO::ColorConfig&
+OCIOColorSystem::colorconfig (ShadingSystemImpl* shadingsys)
+{
+    if (!m_colorconfig) {
+        if (shadingsys)
+            m_colorconfig = shadingsys->colorconfig();
+        else
+            m_colorconfig.reset(new OIIO::ColorConfig);
     }
-#endif
-    return false;
+    return *m_colorconfig;
 }
 
+#endif
 
 
-template <> bool
-ShadingSystemImpl::ocio_transform (StringParam fromspace, StringParam tospace,
-                                   const Dual2<Color3>& C, Dual2<Color3>& Cout) {
-#if OIIO_HAS_COLORPROCESSOR
-    OIIO::ColorProcessorHandle cp;
-    {
-        lock_guard lock (m_mutex);
-        cp = m_ocio_system.load_transform(fromspace, tospace);
-    }
 
-    if (cp) {
-        // Use finite differencing to approximate the derivative. Make 3
-        // color values to convert.
-        const float eps = 0.001f;
-        Color3 CC[3] = { C.val(), C.val() + eps*C.dx(), C.val() + eps*C.dy() };
-        cp->apply ((float *)&CC, 3, 1, 3, sizeof(float), sizeof(Color3), 0);
-        Cout.set (CC[0],
-                  (CC[1] - CC[0]) * (1.0f / eps),
-                  (CC[2] - CC[0]) * (1.0f / eps));
-        return true;
-    }
-#endif
-    return false;
+std::shared_ptr<OIIO::ColorConfig>
+ShadingSystemImpl::colorconfig()
+{
+    lock_guard lock(m_mutex);
```

Comment:
Is the guard only needed when it's not initialized?  Won't this lock a mutex everytime colorconfig is called?

Reply:
The mutex is needed every time ShadingSystemImpl::colorconfig() is called, but it will only be called the first time a ColorConfig is needed by each ShadingContext (that is, each thread) needs a context, after which a copy is retained in the context, where it does not need a lock because it's only used by one thread.

So if there are n threads in your renderer, this lock will be engaged at most n times during the whole render.

Reply:
sounds good, just the name of the method wouldn't make me think it has the potential cost of a lock involved